### PR TITLE
pghrep initial commit

### PR DESCRIPTION
**Pull request by dmius**
Initial commit.
Done:
- check and use options (--checkid, --reportdata(file or data), --outdir)
- loading check data from json file and data from dependencied json files
- use templates from an external file
- simple generation of a report on base of prepared data and store them to an external file
- very simple md template
- using independent plugins for every kind of check
- using tests

TO DO:
- improve md template

### CODE REVIEW

File: pghrep/src/main.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+Perform a generation *md reports on base of results health checks
+Usage: 
+pghrep --checkid=XXX --checkdata=file:///path_to_check_results.json --outdir=/home/results
+*/
+package main
+
+import (
+	"fmt"
+    "os"
```
Comment (NikolayS) [inline pghrep/src/main.go]: the previous line is 1 tab
this one is 4 spaces

in Go programs, the most popular approach is 1 tab (usually visualized as 8 spaces).

let's decide, what will we use?
/cc @Nastradamus

Comment (Nastradamus) [reply pghrep/src/main.go]: We can also use `go fmt` in githook.

File: pghrep/src/main.go [hunk 2]
```diff
@@ -196,53 +194,49 @@ func main() {
...
+        var outputDir string
+        if len(*outDirPtr) == 0 {
+            outputDir = "./"
+        } else {
+            outputDir = *outDirPtr
+        }
+        generateMdReport(reportData, outputDir)
     } else {
-        outputDir = *outDirPtr
+        log.Fatal("Can't generate report on wrong data")
```
Comment (NikolayS) [inline pghrep/src/main.go]: `Cannot generate report on an empty data file.`

File: pghrep/src/pluginsutil.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    pluginsDir string
+    objectsDir string
+    binDir string
+}
+
+func newLoader() (*loader, error) {
+    // The directory that will be watched for new plugins.
+    wd, err := os.Getwd()
+    if err != nil {
+        return nil, fmt.Errorf("could not find current directory: %v", err)
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: `Could..` (as in other error messages)

Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: `Cannot find ..` (as in other error messages)

File: pghrep/src/pluginsutil.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    pluginsDir := filepath.Join(wd, "plugins")
+    binDir := filepath.Join(wd, "bin")
+    //tmp := filepath.Join(wd, "bin")
+
+    // The directory where all .so files will be stored.
+    tmp, err := ioutil.TempDir("", "")
+    if err != nil {
+        return nil, fmt.Errorf("could not create objects dir: %v", err)
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: `Cannot create ...`

File: pghrep/src/pluginsutil.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+}
+
+func (l *loader) destroy() { 
+    //os.RemoveAll(l.objectsDir)
+}
+
+func (l *loader) compileAndRun(name string, data map[string]interface{}) (map[string]interface{}, error) {
+    obj, err := l.compile(name)
+    if err != nil {
+        return nil, fmt.Errorf("could not compile %s: %v", name, err)
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: `Cannot compile ...`

File: pghrep/src/pluginsutil.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    defer os.Remove(obj)
+
+    result, err := l.call(obj, data)
+    if err != nil {
+        return nil, fmt.Errorf("could not run %s: %v", obj, err)
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: `Cannot run`

(or, if you want, let's use `Could...`, bug starting with capital `C`)

File: pghrep/src/pluginsutil.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    return result, nil
+}
+
+// check existance of binary plugin library or compile it from sources
+// and returns its path.
+func (l *loader) get(name string) (string, error) {
+    pluginPath := filepath.Join(l.binDir, name + ".so")
+    _, err := os.Stat(filepath.Join(l.binDir, name + ".so"))
+    if err != nil && os.IsNotExist(err) {
+        log.Println("WARNING: Binary plugin not found. Try compile")
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: let's put plugin name into the error message?

Also: `Try compile` -> `Try compiling` or `Try to compile`

File: pghrep/src/pluginsutil.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        pluginPath, err = l.compile(name)
+    } else {
+        log.Println("Binary plugin found", )
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: why comma in the end?

let's put plugin name into the message?

File: pghrep/src/pluginsutil.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+// plugin, and returns its path.
+func (l *loader) compile(name string) (string, error) {
+    // Copy the file to the objects directory with a different name
+    // each time, to avoid retrieving the cached version.
+    // Apparently the cache key is the path of the file compiled and
+    // there's no way to invalidate it.
+
+    f, err := ioutil.ReadFile(filepath.Join(l.pluginsDir, name + ".go"))
+    if err != nil {
+        return "", fmt.Errorf("could not read %s.go: %v", name, err)
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: `C...`

File: pghrep/src/pluginsutil.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    name = fmt.Sprintf("%d.go", rand.Int())
+    srcPath := filepath.Join(l.objectsDir, name)
+    if err := ioutil.WriteFile(srcPath, f, 0666); err != nil {
+        return "", fmt.Errorf("could not write %s: %v", name, err)
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: `C..`

File: pghrep/src/pluginsutil.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+    return objectPath, nil
+}
+
+// call loads the plugin object in the given path and runs the Run
+// function.
+func (l *loader) call(object string, checkData map[string]interface{}) (map[string]interface{}, error) {
+    p, err := plugin.Open(object)
+    if err != nil {
+        return nil, fmt.Errorf("Check plugin not found.")
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: Let's put details here – what exactly plugin?

File: pghrep/src/pluginsutil.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    symPreparer, err := p.Lookup("Preparer")
+	if err != nil {
+		fmt.Println(err)
+		return nil, fmt.Errorf("plugin failed with error %v", err)
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: Plugin **parsing** failed? Or processing ?

File: pghrep/src/pluginsutil.go [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+	}
+    
+    var preparer Preparer
+	preparer, ok := symPreparer.(Preparer)
+	if !ok {
+        return nil, fmt.Errorf("unexpected type from module symbol")
```
Comment (NikolayS) [inline pghrep/src/pluginsutil.go]: Parsing error, unexpected symbol?

maybe it's worth just to say "Parsing error for plugin XXX"

**Comment by dmius** [comment]
We can merge this branch, but for using we need to develop plugins depended on specified checks.

File: pghrep/templates/report.tpl [hunk 14]
```diff
@@ -2,10 +2,10 @@ Current values
...
 {{.Current}}
 
-Recomended values
+Conclusions
 ===
-{{.Recommended}}
+{{.Conclusion}}
 
-Conclusions
+Recomended values
```
Comment (NikolayS) [inline pghrep/templates/report.tpl]: "Recommended" (double `m`)

Comment (NikolayS) [reply pghrep/templates/report.tpl]: Let's discuss how to name this section better (I had idea that "Recommendations" is the proper subtitle, but definitely we can discuss)

File: pghrep/src/fmtutils/utils.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        return -1
+    } else if (strings.Contains(unit, "EB")) {
+        //factor = 1024 * 1024 * 1024 * 1024 * 1024 * 1024
+        return -1
+    } else if (strings.Contains(unit, "ZB")) {
+        //factor = 1024 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024
+        return -1
+    } else if (strings.Contains(unit, "YB")) {
+        //factor = 1024 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024
+        return -1
```
Comment (NikolayS) [inline pghrep/src/fmtutils/utils.go]: so the results will be ...negative?

File: pghrep/src/fmtutils/utils.go [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+        unit = " PB" // petabyte
+    } else if inputNum >= 1000000000000 {
+        returnVal = RoundUp((inputNum / 1099511627776), precision)
+        unit = " TB" // terrabyte
+    } else if inputNum >= 1000000000 {
+        returnVal = RoundUp((inputNum / 1073741824), precision)
+        unit = " GB" // gigabyte
+    } else if inputNum >= 1000000 {
+        returnVal = RoundUp((inputNum / 1048576), precision)
+        unit = " MB" // megabyte
```
Comment (NikolayS) [inline pghrep/src/fmtutils/utils.go]: let's be more precise here – in Postgres, "MB" means mebibyte, not megabyte. It should be "MiB", but I don't think Postgres will fix the abbreviation soon. In our documents (including comments), let's be precise -- let's use emphasize that it's mebibytes, not megabytes

https://en.wikipedia.org/wiki/Mebibyte
https://en.wikipedia.org/wiki/Megabyte

It applies to other units as well.

File: pghrep/plugins/a003.go [hunk 17]
```diff
@@ -1,15 +1,47 @@
...
-    fmt.Println("Hello from A011!", data)
     result := make(map[string]interface{})
-    result["current"] = "CURRENT VALUES"
+    hosts := pyraconv.ToStringArray(data["hosts"])
+    checkData := data["checkData"].(map[string]interface{})
+    for _, host := range hosts {
+        fmt.Println("Host:", host)
+        hostSettings := checkData[host].([]interface{})
+        shared_buffers := getSharefBuffersValue(hostSettings)
+        result["current"] = "Current value of settings shared_buffers is " + fmtutils.ByteFormat(float64(shared_buffers), 0)
```
Comment (NikolayS) [inline pghrep/plugins/a003.go]: as discussed, let's use less words. Like:
```
Shared buffers: VALUE
```

File: pghrep/plugins/a003.go [hunk 18]
```diff
@@ -1,15 +1,47 @@
...
 type prepare string
 
+func getSharefBuffersValue(hostSettings []interface{}) int64 {
+    for _, setting := range hostSettings {
+        curSetting := setting.(map[string]interface{})
+        if curSetting["name"] == "shared_buffers" {
+            if curSetting["unit"] != nil {
+                val, err := strconv.ParseInt(pyraconv.ToString(curSetting["setting"]), 10, 64)
+                if err != nil {
+                    return -1
```
Comment (NikolayS) [inline pghrep/plugins/a003.go]: does `-1` mean something particular?

If we return `-1`, what will happen next? An error?

File: pghrep/plugins/a003.go [hunk 19]
```diff
@@ -1,15 +1,47 @@
...
+    "../src/pyraconv"
+    "../src/fmtutils"
+    "strconv"
+)
 
 var Data map[string]interface{}
 
 type prepare string
 
+func getSharefBuffersValue(hostSettings []interface{}) int64 {
```
Comment (NikolayS) [inline pghrep/plugins/a003.go]: a typo here: "Sharef"

File: pghrep/plugins/a003.go (hunk 18 above)
Comment (NikolayS) [reply pghrep/plugins/a003.go]: BTW, why not `return nil` ?

File: pghrep/src/fmtutils/utils.go [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+     round = math.Ceil(digit)
+     newVal = round / pow
+     return
+}
+
+func ByteFormat(inputNum float64, precision int) string {
+    var unit string
+    var returnVal float64
+
+    if inputNum >= 1000000000000000000000000 {
```
Comment (NikolayS) [inline pghrep/src/fmtutils/utils.go]: isn't it better to write `     if inputNum >= 100^26  {` here?

File: pghrep/src/fmtutils/utils.go [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+        returnVal = RoundUp((inputNum / 1208925819614629174706176), precision)
+        unit = " YB" // yottabyte
```
Comment (NikolayS) [inline pghrep/src/fmtutils/utils.go]: yobibyte :)

File: pghrep/templates/report.tpl (hunk 14 above)
Comment (Nastradamus) [reply pghrep/templates/report.tpl]: Agree with @NikolayS

File: pghrep/plugins/a003.go (hunk 18 above)
Comment (dmius) [reply pghrep/plugins/a003.go]: `plugins/a003.go:21:21: cannot use nil as type int64 in return argument`

**Comment by Nastradamus** [comment]
:tada:

**Pull request merged**